engine/table/src/main/java/io/deephaven/engine/table/impl/asofjoin/StaticAsOfJoinStateManagerTypedBase.java

```diff
@@ -0,0 +1,390 @@
+        leftRowSetSource = new ImmutableObjectArraySource<>(RowSetBuilderSequential.class, null);
+        rightRowSetSource = new ImmutableObjectArraySource<>(Object.class, null);
+        // endregion constructor
+
+        for (int ii = 0; ii < tableKeySources.length; ++ii) {
+            chunkTypes[ii] = keySourcesForErrorMessages[ii].getChunkType();
+            mainKeySources[ii] = InMemoryColumnSource.getImmutableMemoryColumnSource(tableSize,
+                    tableKeySources[ii].getType(), tableKeySources[ii].getComponentType());
+        }
+
+        this.maximumLoadFactor = maximumLoadFactor;
+
+        ensureCapacity(tableSize);
+    }
+
+    private void ensureCapacity(int tableSize) {
+        for (WritableColumnSource<?> mainKeySource : mainKeySources) {
+            mainKeySource.ensureCapacity(tableSize);
+        }
+        // region ensureCapacity
+        leftRowSetSource.ensureCapacity(tableSize);
+        rightRowSetSource.ensureCapacity(tableSize);
+        // endregion ensureCapacity
+    }
+
+    BuildContext makeBuildContext(ColumnSource<?>[] buildSources, long maxSize) {
+        return new BuildContext(buildSources, (int) Math.min(CHUNK_SIZE, maxSize));
+    }
+
+    ProbeContext makeProbeContext(ColumnSource<?>[] buildSources, long maxSize) {
+        return new ProbeContext(buildSources, (int) Math.min(CHUNK_SIZE, maxSize));
+    }
+
+    static boolean addIndex(ImmutableObjectArraySource<RowSetBuilderSequential> source, long location, long keyToAdd) {
+        boolean addedSlot = false;
+        RowSetBuilderSequential builder = source.getUnsafe(location);
+        if (builder == null) {
+            source.set(location, builder = RowSetFactory.builderSequential());
+            addedSlot = true;
+        }
+        builder.appendKey(keyToAdd);
+        return addedSlot;
+    }
+
+    /**
+     * Returns true if this is the first left row key added to this slot.
+     */
+    protected boolean addLeftIndex(long tableLocation, long keyToAdd) {
+        return addIndex(leftRowSetSource, tableLocation, keyToAdd);
+    }
+
+    protected void addRightIndex(long tableLocation, long keyToAdd) {
+        // noinspection unchecked
+        addIndex((ImmutableObjectArraySource) rightRowSetSource, tableLocation, keyToAdd);
+    }
+
+    protected void buildTable(
+            final BuildContext bc,
+            final RowSequence buildRows,
+            final ColumnSource<?>[] buildSources,
+            final TypedHasherUtil.BuildHandler buildHandler) {
+        try (final RowSequence.Iterator rsIt = buildRows.getRowSequenceIterator()) {
+            // noinspection unchecked
+            final Chunk<Values>[] sourceKeyChunks = new Chunk[buildSources.length];
+
+            while (rsIt.hasMore()) {
+                final RowSequence chunkOk = rsIt.getNextRowSequenceWithLength(bc.chunkSize);
+                final int nextChunkSize = chunkOk.intSize();
+
+                if (exceedsCapacity(nextChunkSize)) {
```

Comment:
I think this is going to be incorrect.  We want to allow rehashing for aj, because we have no idea how many buckets will actually exist.  we might have only a few dozen buckets in a 100MM row table.  naturalJoin has different behavior because it is constrained to be 1-1 on the right side